numpy/core/src/umath/ufunc_object.c

```diff
@@ -4169,15 +4172,23 @@ PyUFunc_GenericReduction(PyUFuncObject *ufunc, PyObject *args,
         return (PyObject *)ret;
     }
 
-    if (Py_TYPE(op) != Py_TYPE(ret)) {
-        res = PyObject_CallMethod(op, "__array_wrap__", "O", ret);
-        if (res == NULL) {
-            PyErr_Clear();
-        }
-        else if (res == Py_None) {
-            Py_DECREF(res);
+    else {
+        PyObject *wrap;
+        _find_array_wrap(args, kwds, &wrap, 1, 1);
+
+        if (wrap == Py_None) {
+            Py_DECREF(wrap);
         }
-        else {
+        else if (wrap != NULL) {
+            PyObject *res = PyObject_CallFunctionObjArgs(
```

Comment:
Would almost prefer putting the initialization fo `res` on another line just to have space to put it all on a single line.